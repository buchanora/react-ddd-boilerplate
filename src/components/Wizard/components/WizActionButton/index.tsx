import React, {PropsWithChildren} from "react";
import {Button} from 'components/Button';
import {ActionIcon} from './Icon';

interface Props extends InheritedProps{
  direction: Direction;
  getActionState?(step?: number): ActionState;
  text: string;
  primary?: boolean;
  onClick?(direction: Direction): boolean;
}

interface InheritedProps {
  activeView?: number;
  goBack?(): void;
  moveTo?(step: number): void;
}

interface ActionState {
  disabled: boolean;
  visible: boolean;
  label: string;
}

export type WizActionButtonState = ActionState;

export type Direction = 'backward' | 'forward';

export class WizActionButton extends React.Component<Props> {
  static displayName = "WizActionButton";

  constructor(props: PropsWithChildren<Props>){
    super(props);
  }

  render() {

    function actionStateClasses(_actionState: ActionState): string {
      let classes = "";
      if (_actionState.disabled) {
        classes += " wizard-action--disabled";
      }
      if (!_actionState.visible) {
        classes += " wizard-action--hidden";
      }
      return classes;
    }

    const actionState: ActionState = this.props.getActionState
      ? this.props.getActionState(this.props.activeView)
      : {
        disabled: false,
        label: this.props.text,
        visible: true
      };

    return (
      <Button
        iconLeft={this.props.direction === "backward" ? 'arrow-right' : undefined}
        iconRight={this.props.direction === "forward" ? 'arrow-left' : undefined}
        text={actionState.label}
        size='large'
        onClick={
          actionState && actionState.disabled
            ? () => {}
            : () => this.props.onClick && this.props.onClick(this.props.direction)
        }
        style={this.props.direction === 'backward' ? 'hollow' : 'solid'}
        disabled={actionState.disabled}
        color={'primary'}
        className={`${!actionState.visible ? 'wizard-action--hidden' : ''}`}>
      </Button>
    );
  }
}

export default WizActionButton;
