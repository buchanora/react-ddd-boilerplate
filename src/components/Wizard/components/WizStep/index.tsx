import React, {PropsWithChildren} from "react";

interface Props {
  label?: string;
  index?: number;
  last?: boolean;
  first?: boolean;
  count?: number;
  isActive?: boolean;
  isDone?: boolean;
}

export class WizStep extends React.Component<Props> {
  static displayName = 'WizStep';
  static defaultProps = {
    label: 'Step'
  };

  constructor(props: PropsWithChildren<Props>){
    super(props);
  }

  render(){
    const index = this.props.index || 0;
    const isLastStep = this.props.last ? this.props.last : this.props.count === index + 1;
    const isFirstStep = this.props.first ? this.props.first : this.props.index === 0;

    return (
      <span
        className={`wizard-step ${isLastStep ? "wizard-step--last" : ""} ${
          isFirstStep ? "wizard-step--first" : ""
        } ${this.props.isDone ? "wizard-step--done" : ""} ${
          this.props.isActive ? "wizard-step--active" : ""
        }`}
      >
        {isFirstStep || <span className="wizard-step-divider" />}

        <span className={`wizard-step-countWrapper`}>
          {<span className="wizard-step-count">{index + 1 || 1}</span>}
        </span>
        {this.props.label && <span className="wizard-step-label">{this.props.label}</span>}
      </span>
    );
  }
}

export default WizStep;
