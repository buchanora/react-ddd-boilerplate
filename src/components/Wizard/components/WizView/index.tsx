import React, { Component, PropsWithChildren } from "react";

interface Props {
  index?: number;
}

export class WizView extends Component<Props> {
  static displayName: string = 'WizView';

  constructor(props: PropsWithChildren<Props>) {
    super(props);
  }

  render() {
    const { children } = this.props;
    return <div className={`wizard-body`}>{children}</div>;
  }
}

export default WizView;
