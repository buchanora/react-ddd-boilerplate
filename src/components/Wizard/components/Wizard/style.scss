@use "theme" ;
.wizard-wrapper {
    background-color: inherit;
    display: flex;
    flex: 1;
    flex-direction: row;
    align-items: stretch;
    justify-content: space-between;
    height: 100%;
    width: 100%;
    flex-wrap: wrap;
    @include theme.breakpoint(medium){
        flex-direction: column;
    }
}
.submitTrigger{
    display: none;
}
