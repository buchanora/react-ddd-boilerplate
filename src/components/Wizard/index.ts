export * from './components/Wizard';
export * from './components/WizStep';
export * from './components/WizSteps';
export * from './components/WizView';
export * from './components/WizViews';
export *  from './components/WizActionButton';
export * from './components/WizActions';
