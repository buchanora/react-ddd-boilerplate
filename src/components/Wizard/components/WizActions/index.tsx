import React, {ReactNode, PropsWithChildren} from 'react';
import WizActionButton, {Direction} from '../WizActionButton';
import {ButtonGroup} from 'components/Button';

interface Props {
  activeView?: number;
  goBack?(): void;
  goForward?(): void;
  moveTo?(step: number): void;
  render?(props: Omit<Props, 'render'>): ReactNode;
}

export class WizActions extends React.Component<Props> {
  static displayName = 'WizActions';

  constructor(props: PropsWithChildren<Props>){
    super(props);
  }

  render() {

    let computedChildren: ReactNode = [];

    if(this.props.render && typeof this.props.render === 'function') {
      computedChildren = this.props.render({
        activeView: this.props.activeView,
        moveTo: this.props.moveTo,
        goBack: this.props.goBack
      });
    } else {
      computedChildren = React.Children.map(this.props.children, (child, index) => {
        if(React.isValidElement(child)) {
          if(typeof child.type === 'string') {
            return child;
          }
          if(child.type instanceof WizActionButton || child.type.name === 'WizActionButton'){

            const childProps = {
              key: 'action' + index,
              activeView: this.props.activeView,
              text: child.props.text,
              direction: child.props.direction,
              onClick: (direction: Direction) => this._handleActionClick(direction),
              getActionState: child.props.getActionState
                ? child.props.getActionState
                : () => ({
                  disabled: false,
                  visible: child.props.direction === 'backward' && this.props.activeView === 0 && index === 0 ? false :true,
                  label: child.props.text
                })
            }

            return React.cloneElement(child, childProps, child.props.children);
          }
          return React.cloneElement(child, {key: 'action' + index}, child.props.children);
        }
      });
    }
    return (
      <footer className='wizard-actions'>
        <ButtonGroup>
        {computedChildren}
        </ButtonGroup>
      </footer>
    );
  }
  _handleActionClick(direction: string) {
    if(direction === 'forward') {
      this.props.goForward && typeof this.props.goForward === 'function' && this.props.goForward();
    }
    if(direction === 'backward'){
      this.props.goBack && typeof this.props.goBack === 'function' && this.props.goBack();
    }
  }
}

export default WizActions;
