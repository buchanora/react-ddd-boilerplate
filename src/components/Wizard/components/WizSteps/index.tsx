import React, {PropsWithChildren} from "react";
import WizStep from "../WizStep";

interface Props {
  activeView?: number;
  done?: Array<number>;
  render?(props: Omit<Props, "render">): React.ReactNode;
}

export class WizSteps extends React.Component<Props> {
  constructor(props: PropsWithChildren<Props>){
    super(props);
  }

  static displayName = "WizSteps";
  static defaultProps = {
    activeView: 0,
    done: [],
  };
  render() {
    const { children, activeView, done, render } = this.props;
    let steps;

    if(render && typeof this.props.render === "function") {
      steps = render({ activeView, done });
    } else {
      steps = React.Children.toArray(children)
        .filter((child) => {
          if(
            React.isValidElement(child) &&
            typeof child.type !== "string" &&
            typeof child.type !== null &&
            (child.type instanceof WizStep || child.type.name === 'WizStep')
          ) {
            return child;
          }
        })
        .map((child, index, childArray) => {
          if(React.isValidElement(child)) {
            return React.cloneElement(child, {
              key: "step_" + index,
              index: index,
              label: child.props.label,
              isDone: done && done[index] ? true : false,
              isActive: activeView === index,
              count: childArray.length,
            },
            child.props.children);
          }
        });
    }
    return <header className="wizard-steps">{steps}</header>;
  }
}
export default WizSteps;
