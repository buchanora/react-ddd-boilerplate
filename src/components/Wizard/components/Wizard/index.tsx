import React, { Component, PropsWithChildren } from "react";
import {Direction} from '../WizActionButton';
import {WizActions} from "../WizActions";
import {WizSteps} from '../WizSteps';
import {WizViews} from '../WizViews';

type Props = {
  active: number;
  as: "div" | "form";
  render?(props: {}): React.ReactNode;
  onNext?(val: number): void;
  onPrev?(val: number): void;
  loading?: boolean;
};

type State = {
  activeView: number;
  done: boolean[];
  action: Direction;
};

export class Wizard extends Component<Props, State> {
  private _viewCount: number;
  static displayName: string = 'Wizard';
  static defaultProps = {
    active: 0,
    as: "div",
    displayName: "Wizard",
  };
  state: State = {
    activeView: 0,
    done: [],
    action: 'forward'
  }
  constructor(props: PropsWithChildren<Props>) {
    super(props);
    this._viewCount = 0;
  }
  public render() {
    const { as, render, loading } = this.props;
    const { activeView, done } = this.state;

    const moveTo = (viewIndex: number) => this._handleMoveTo(viewIndex);
    const goBack = () => this._handleBackClick();
    const goForward = () => this._handleForwardClick();
    const markAsDone = (viewIndex: number) => this._handleDone(viewIndex);

    let children;
    if (render) {
      children = render({
        activeView,
        moveTo,
        goForward,
        goBack,
        loading,
        done,
        markAsDone,
      });
    } else {
      children = React.Children.toArray(this.props.children)
        .filter((child) => {
          if (
            React.isValidElement(child) &&
            typeof child.type !== "string" &&
            (
              (child.type instanceof WizSteps || child.type.name === 'WizSteps') ||
              (child.type instanceof WizViews || child.type.name === 'WizViews') ||
              (child.type instanceof WizActions || child.type.name === 'WizActions')
            )) {
              return child;
          }
        })
        .map((child, index, all) => {
          if (React.isValidElement(child) && typeof child.type !== "string") {
            if (child.type instanceof  WizSteps || child.type.name === 'WizSteps') {
              return React.cloneElement(child, {
                key: "wiz-steps" + index,
                activeView,
                loading,
                done,
              },
              child.props.children);
            }
            if(child.type instanceof WizViews || child.type.name === 'WizViews') {
              this._viewCount = React.Children.count(child.props.children);
              return React.cloneElement(child, {
                key: "wiz-views" + index,
                getAction: () => this.state.action,
                activeView,
                moveTo,
                goBack,
                loading,
              },
              child.props.children);
            }
            if(child.type instanceof WizActions || child.type.name === 'WizActions') {
              return React.cloneElement(
                child,
                {
                  key: "wiz-actions" + index,
                  activeView,
                  moveTo,
                  goBack,
                  goForward,
                  loading,
                },
                child.props.children
              );
            }
          }
        });
    }

    return as === "form" ? (
      <form className="wizard-wrapper" onSubmit={(e) => {
          e.preventDefault();
          this._handleForwardClick();
        }}>
        {children}
        <input type="submit" className="submitTrigger" />
      </form>
    ) : (
      <div className="wizard-wrapper">{children}</div>
    );
  }
  _handleBackClick() {
    this.setState(() => ({action: 'backward'}));
    setTimeout(() => {
      this.setState((curState) => {
        let activeView = curState.activeView;
        if(curState.activeView !== 0) {
          activeView = curState.activeView - 1
        }
        this.props.onPrev && this.props.onPrev(activeView);
        return {
          activeView,
        }
      });

    }, 100);
  }

  _handleForwardClick() {
    this.setState(() => ({action: 'forward'}));
    setTimeout(() => {
      this.setState((cur) => {
        let activeView = cur.activeView + 1;
        if(activeView >= this._viewCount) {
          activeView = this._viewCount - 1;
        }
        this.props.onNext && this.props.onNext(activeView);
        return {
          activeView
        };
      });
    }, 100);
  }

  _handleMoveTo(activeView: number) {
    this.setState((cur) => {
      const nowActive = !this.props.render
        ? activeView <= this._viewCount
          ? activeView
          : this._viewCount
        : activeView;
      if (cur.activeView < nowActive) {
        this.props.onNext && this.props.onNext(nowActive);
      } else {
        this.props.onPrev && this.props.onPrev(nowActive);
      }
      return {
        activeView: nowActive,
        action: cur.activeView < nowActive ? "forward" : "backward",
      };
    });
  }

  _handleDone(index: number) {
    const done: boolean[] = [];
    this.setState((cur) => {
      done
        .fill(false, 0, index) //can only be filled with the array type value
        .map((d, i) => (i === index ? true : cur.done[i]));
      return { done };
    });
  }
}

export default Wizard;
