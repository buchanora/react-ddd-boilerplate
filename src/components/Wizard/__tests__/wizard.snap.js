import React from 'react';
import renderer from 'react-test-renderer';
import {Wizard, WizSteps, WizStep, WizViews, WizView, WizActions, WizActionButton} from 'components/Wizard';

test('Wizard renders the right children', () => {
    const component = renderer.create(
      <Wizard >
        <WizSteps>
            <WizStep label='1'></WizStep>
            <WizStep label='2'></WizStep>
        </WizSteps>
        <WizViews>
            <WizView>1</WizView>
            <WizView>2</WizView>
        </WizViews>
        <WizActions>
            <WizActionButton />
            <WizActionButton text='Save' />
            <WizActionButton  primary/>
        </WizActions>
      </Wizard>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });