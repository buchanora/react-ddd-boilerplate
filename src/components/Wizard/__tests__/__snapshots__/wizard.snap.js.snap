// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wizard renders the right children 1`] = `
<div
  className="wizard-wrapper"
>
  <header
    className="wizard-steps"
  >
    <span
      className="wizard-step  wizard-step--first  wizard-step--active"
    >
      <span
        className="wizard-step-countWrapper"
      >
        <span
          className="wizard-step-count"
        >
          1
        </span>
      </span>
      <span
        className="wizard-step-label"
      >
        1
      </span>
    </span>
    <span
      className="wizard-step wizard-step--last   "
    >
      <span
        className="wizard-step-divider"
      />
      <span
        className="wizard-step-countWrapper"
      >
        <span
          className="wizard-step-count"
        >
          2
        </span>
      </span>
      <span
        className="wizard-step-label"
      >
        2
      </span>
    </span>
  </header>
  <div
    className="wizard-viewWrapper"
  >
    <div
      className="wizard-body"
    >
      1
    </div>
  </div>
  <footer
    className="wizard-actions"
  >
    <div
      className="buttonGroup__wrapper  "
    >
      <button
        className="solid solid-primary button  button--large  "
        onClick={[Function]}
        type="button"
      />
      <button
        className="solid solid-primary button  button--large  "
        onClick={[Function]}
        type="button"
      >
        Save
      </button>
      <button
        className="solid solid-primary button  button--large  "
        onClick={[Function]}
        type="button"
      />
    </div>
  </footer>
</div>
`;
